### PR TITLE
(feat) - warn about hooks outside render and/or components

### PR DESCRIPTION
**Pull request by JoviDeCroock**
Adds: 11B to main package (this seems a bit much for one if check ?)

### CODE REVIEW

**Comment by JoviDeCroock** [comment]
coverall is seeing ghosts

File: src/diff/index.js [hunk 1]
```diff
@@ -134,6 +134,8 @@ export function diff(dom, parentDom, newVNode, oldVNode, context, isSvg, excessD
 			let vnode = c._prevVNode = coerceToVNode(c.render(c.props, c.state, c.context));
 			c._dirty = false;
 
+			if (options.afterRender) options.afterRender();
```
Comment (marvinhagemeister) [inline src/diff/index.js]: Wondering if we could use `options.diffed` instead. It's callsite is just a couple lines later at the end of `diff()`.

File: hooks/src/index.js [hunk 2]
```diff
@@ -21,6 +21,12 @@ options.render = vnode => {
 };
 
 
+let oldDiffed = options.diffed;
+options.diffed = () => {
+	if (oldDiffed) oldDiffed();
+	currentComponent = undefined;
+};
+
```
Comment (robertknight) [inline hooks/src/index.js]: There is another `options.diffed` hook just below. Any reason not to fold the `currentComponent` assignment into that?

File: debug/test/browser/debug.test.js [hunk 3]
```diff
@@ -84,7 +104,7 @@ describe('debug', () => {
 				<p>{state}</p>
 			);
 		};
-		const fn = () => render(<App />, scratch);
+		const fn = () => act(() => render(<App />, scratch));
```
Comment (robertknight) [inline debug/test/browser/debug.test.js]: What is the effect of adding the `act` call here?

File: src/index.d.ts [hunk 4]
```diff
@@ -191,6 +191,8 @@ declare namespace preact {
 		diff?(vnode: VNode): void;
 		/** Attach a hook that is invoked before a vnode has rendered. */
 		render?(vnode: VNode): void;
+		/** Attach a hook that is invoked before a hook's state is queried. */
+		preHook?(component: VNode): void;
```
Comment (robertknight) [inline src/index.d.ts]:
```suggestion
		preHook?(component: Component): void;
```

The type of the argument here isn't a `VNode`.

File: debug/test/browser/debug.test.js (hunk 3 above)
Comment (JoviDeCroock) [reply debug/test/browser/debug.test.js]: Hooks inside the render method of App will be flushed

File: hooks/src/index.js (hunk 2 above)
Comment (JoviDeCroock) [reply hooks/src/index.js]: Thought it would improve readability a bit, I'll move it

File: debug/src/debug.js [hunk 5]
```diff
@@ -101,6 +101,12 @@ export function initDebug() {
 		if (oldBeforeDiff) oldBeforeDiff(vnode);
 	};
 
+	options.preHook = (comp) => {
```
Comment (marvinhagemeister) [inline debug/src/debug.js]: Changing `preHook` to just `hook` saves `3 B` :tada:

File: hooks/src/index.js [hunk 6]
```diff
@@ -34,6 +33,7 @@ options.diffed = vnode => {
 	// TODO: Consider moving to a global queue. May need to move
 	// this to the `commit` option
 	hooks._pendingLayoutEffects = handleEffects(hooks._pendingLayoutEffects);
+	currentComponent = undefined;
```
Comment (marvinhagemeister) [inline hooks/src/index.js]: Removing this line has no effect on the test outcome :+1:

**Pull request merged**